### PR TITLE
Extend Moshi DTOs

### PR DESCRIPTION
**Pull request by zsmb13**
### Description

Moshi adventures, pt 3.

Adds all fields for Message DTOs, adding other necessary DTOs, new mapping code, and some bugfixes / code reorg along the way.

---

Planned next PR (and therefore not handled here yet):

- Expanding the User object
- Support for the custom object / extra data behaviour on User, Attachment, Reaction
- Sorting the fields within the DTOs by some logic (currently scrambled)

### Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [ ] ~Changelog updated with client-facing changes~
- [x] New code is covered by unit tests
- [ ] ~Comparison screenshots added for visual changes~
- [x] Reviewers added

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/parser2/CustomObjectDtoAdapter.kt [hunk 1]
```diff
@@ -68,10 +68,15 @@ internal abstract class CustomObjectDtoAdapter<Value : Any>(private val kClass:
...
         jsonWriter: JsonWriter,
-        value: Value,
+        value: Value?,
         mapAdapter: JsonAdapter<MutableMap<String, Any?>>,
         valueAdapter: JsonAdapter<Value>,
     ) {
+        if (value == null) {
+            jsonWriter.nullValue()
+            return
+        }
```
Comment (zsmb13) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/parser2/CustomObjectDtoAdapter.kt:78]: This was an unhandled case before, and caused a crash in tests.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/DtoMapper.kt [hunk 2]
```diff
@@ -1,60 +0,0 @@
-package io.getstream.chat.android.client.api2
-
-import io.getstream.chat.android.client.api2.model.dto.DownstreamMessageDto
-import io.getstream.chat.android.client.api2.model.dto.UpstreamMessageDto
-import io.getstream.chat.android.client.api2.model.dto.UserDto
-import io.getstream.chat.android.client.models.Message
-import io.getstream.chat.android.client.models.User
-
-internal class DtoMapper {
```
Comment (zsmb13) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/DtoMapper.kt:9]: Split out into several files as extension functions

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/mapping/MessageMapping.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        text = text,
+        html = html,
+        parentId = parent_id,
+        command = command,
+        user = user.toDomain(),
+        replyCount = reply_count,
+        reactionCounts = reaction_counts.toMutableMap(),
+        reactionScores = reaction_scores.toMutableMap(),
+        latestReactions = latest_reactions.mapTo(mutableListOf(), ReactionDto::toDomain),
+        ownReactions = own_reactions.mapTo(mutableListOf(), ReactionDto::toDomain),
```
Comment (zsmb13) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/mapping/MessageMapping.kt:50]: Some of the `MutableMap` and `MutableList` types in the domain models are really ugly, so we should eventually change them. For the time being though, we can deal with them just fine here with a bit of extra code.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/model/dto/UpstreamMessageDto.kt [hunk 4]
```diff
@@ -1,17 +1,35 @@
 package io.getstream.chat.android.client.api2.model.dto
 
 import com.squareup.moshi.JsonClass
+import java.util.Date
 
 @JsonClass(generateAdapter = true)
 internal data class UpstreamMessageDto(
     val id: String,
     val cid: String,
+    val parent_id: String?,
```
Comment (leandroBorgesFerreira) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/model/dto/UpstreamMessageDto.kt:10]: I think that adding @Json(name = "[field name]") to all fields woudl be nice, so we prevent proguard problems. Just making sure that the `[...]/api2/model/dto` is in the proguard rules works too.

**Review by leandroBorgesFerreira — APPROVED** [review]

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/model/dto/UpstreamMessageDto.kt (hunk 4 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/model/dto/UpstreamMessageDto.kt:10]: I prefer doing both =P

Comment (zsmb13) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/model/dto/UpstreamMessageDto.kt:10]: We don't have these annotations on purpose, see https://github.com/GetStream/stream-chat-android/pull/1274#discussion_r564593801

As for Proguard: Moshi's generated adapters contain the string names of the fields as strings, so obfuscation shouldn't affect them at all.

Here's an example of the trie that Moshi uses for figuring out which field to parse being created in the generated code: 
image

Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/api2/model/dto/UpstreamMessageDto.kt:10]: Makes sense. Solving the conversation.

**Review by adasiewiczr — APPROVED** [review]

**Pull request merged**